ext/opentelemetry-ext-asgi/src/opentelemetry/ext/asgi/__init__.py

```diff
@@ -0,0 +1,206 @@
+            return StatusCanonicalCode.DEADLINE_EXCEEDED
+        return StatusCanonicalCode.INTERNAL
+    return StatusCanonicalCode.UNKNOWN
+
+
+def collect_request_attributes(scope):
+    """Collects HTTP request attributes from the ASGI scope and returns a
+    dictionary to be used as span creation attributes."""
+
+    port = scope.get("server")[1]
+    server_host = scope.get("server")[0] + (
+        ":" + str(port) if port != 80 else ""
+    )
+    http_url = scope.get("scheme") + "://" + server_host + scope.get("path")
+    if scope.get("query_string"):
+        http_url = http_url + ("?" + scope.get("query_string").decode("utf8"))
+
+    result = {
+        "component": scope.get("type"),
+        "http.method": scope.get("method"),
+        "http.server_name": scope.get("server")[0],
+        "http.scheme": scope.get("scheme"),
+        "http.host": server_host,
+        "host.port": port,
+        "http.flavor": scope.get("http_version"),
+        "http.target": scope.get("path"),
+        "http.url": http_url,
+    }
+
+    if "client" in scope:
+        result["net.peer.ip"] = scope.get("client")[0]
+        result["net.peer.port"] = scope.get("client")[1]
+
+    return result
+
+
+def set_status_code(span, status_code):
+    """Adds HTTP response attributes to span using the status_code argument."""
+    try:
+        status_code = int(status_code)
+    except ValueError:
+        span.set_status(
+            Status(
+                StatusCanonicalCode.UNKNOWN,
+                "Non-integer HTTP status: " + repr(status_code),
+            )
+        )
+    else:
+        span.set_attribute("http.status_code", status_code)
+        span.set_status(Status(http_status_to_canonical_code(status_code)))
+
+
+def get_default_span_name(scope):
+    """Calculates a (generic) span name for an incoming HTTP request based on the ASGI scope."""
+
+    # TODO: Update once
+    #  https://github.com/open-telemetry/opentelemetry-specification/issues/270
+    #  is resolved
+    return scope.get("path", "/")
+
+
+class OpenTelemetryMiddleware:
+    """The ASGI application middleware.
+
+    This class is an ASGI middleware that starts and annotates spans for any
+    requests it is invoked with.
+
+    Args:
+        app: The ASGI application callable to forward requests to.
+    """
+
+    def __init__(self, app):
+        self.app = guarantee_single_callable(app)
+        self.tracer = trace.tracer_source().get_tracer(__name__, __version__)
+
+    async def __call__(self, scope, receive, send):
+        """The ASGI application
+
+        Args:
+            scope: A ASGI environment.
+            receive: An awaitable callable yielding dictionaries
+            send: An awaitable callable taking a single dictionary as argument.
+        """
+
+        parent_span = propagators.extract(get_header_from_scope, scope)
+        span_name = get_default_span_name(scope)
+
+        with self.tracer.start_as_current_span(
```

Comment:
Is the intention to create a span for the `lifespan` [event](https://asgi.readthedocs.io/en/latest/specs/lifespan.html#scope)? Both `uvicorn` and `hypercorn` implement the `lifespan` protocol

Reply:
No, that was not intentional. So far I only tested with `daphne`. I can add a short-circuit for lifespan events, if you think that's preferable?

Reply:
I believe so. I can't think of a reason someone would want to include a trace span for the "lifespan" ASGI event in their HTTP server-centric tracing.

A pattern I've seen in other ASGI middleware is:
```
async def __call__(self, scope, receive, send):
    if scope["type"] not in {"http", "ws"}:
            return await self.app(scope, receive, send)
    # ...do trace otherwise
```